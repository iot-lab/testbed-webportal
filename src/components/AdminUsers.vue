<template>
  <div class="container mt-3">

    <h2>Users</h2>
    <div class="row">
      <div class="col-md-4 mb-4">
        <div class="my-2">
          <span class="lead mr-3">Show</span>                
          <label class="custom-control custom-radio">
            <input id="radioStacked1" name="radio-stacked" type="radio" class="custom-control-input" v-model="show" value="pending" @click="showPending">
            <span class="custom-control-indicator"></span>
            <span class="custom-control-description">Pending users</span>
          </label>
          <label class="custom-control custom-radio mr-0">
            <input id="radioStacked2" name="radio-stacked" type="radio" class="custom-control-input" v-model="show" value="admin" @click="showAdmin">
            <span class="custom-control-indicator"></span>
            <span class="custom-control-description">Administrators</span>
          </label>
        </div>
        <form @submit.prevent="search">
          <div class="input-group">
            <input type="text" class="form-control" placeholder="Search for users" v-model="pattern">
            <div class="input-group-btn">
              <button type="submit" class="btn btn-secondary" aria-label="Search"> <i class="fa fa-search"></i> </button>
            </div>
          </div>
        </form>
      </div>
      <div class="col-md-8 text-right align-self-end mb-4">
        <router-link :to="{name:'addUsers'}" class="btn btn-success"><i class="fa fa-user-plus"></i> Add Users</router-link>
      </div>
    </div>
    <table class="table table-striped table-sm">
      <thead>
        <tr>
          <th class="header headerSortDown" aria-sort="ascending">Login</th>
          <th class="header">First name</th>
          <th class="header">Last name</th>
          <th class="header">Email</th>
          <th class="header">Created</th>
          <th class="header" width="50px">State</th>
          <th class="header" width="50px">Role</th>
          <th class="header" width="50px">Actions</th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="(user, i) in users">
          <td>{{user.login}}</td>
          <td>{{user.firstName}}</td>
          <td>{{user.lastName}}</td>
          <td><a :href="'mailto:' + user.email" class="email">{{user.email}}</a></td>
          <td :title="user.created | formatDateTime">{{user.created | formatDate}}</td>
          <td>
            <a href="" class="btn btn-sm" style="width: 70px"
              :class="(user.status === 'pending') ? 'btn-primary' : 'btn-outline-primary'"
              v-tooltip:top="(user.status === 'pending') ? 'Activate' : 'Deactivate'"
              @click.prevent="toggleActive(user)">
              {{user.status}}
            </a>
          </td>
          <td>
            <a href="" class="btn btn-sm" style="width: 70px"
              :class="user.admin ? 'btn-warning' : 'btn-outline-primary'"
              v-tooltip:top="user.admin ? ' User' : 'Admin'"
              @click.prevent="toggleAdmin(user)">
              {{user.admin ? 'Admin': 'User'}}
            </a>
          </td>
          <td>
            <div class="btn-group" role="group" aria-label="User actions">
              <a href="" class="btn btn-sm border-0 btn-outline-secondary" v-tooltip="'Experiments'" @click.prevent="">
                <i class="fa fa-fw fa-flask"></i>
              </a>
              <a href="" class="btn btn-sm border-0 btn-outline-secondary" v-tooltip="'Edit'"
                data-toggle="modal" data-target=".edit-user-modal" @click="currentUser = user; $refs.user.clean(user)">
                <i class="fa fa-fw fa-pencil"></i>
              </a>
              <a href="" class="btn btn-sm border-0 btn-outline-secondary" v-tooltip="'Email'"
                data-toggle="modal" data-target=".email-user-modal" @click="currentUser = user; $refs.mail.$validator.clean()">
                <i class="fa fa-fw fa-envelope"></i>
              </a>
              <a href="" class="btn btn-sm border-0 btn-outline-danger" v-tooltip="'Reset password'" @click.prevent="resetPassword(user)">
                <i class="fa fa-fw fa-unlock-alt"></i>
              </a>
              <a href="" class="btn btn-sm border-0 btn-outline-danger" v-tooltip="'Delete'" @click.prevent="deleteUser(user)">
                <i class="fa fa-fw fa-trash"></i>
              </a>
            </div>
          </td>
        </tr>
      </tbody>
    </table>

    <div class="text-muted">
      <p v-if="users && users.length == 1">{{users.length}} matching user</p>
      <p v-else-if="users && users.length > 1">{{users.length}} matching users</p>
      <p v-else>No matching user found</p>        
    </div>

    <div class="modal fade edit-user-modal" tabindex="-1" role="dialog" aria-labelledby="editUserModal" aria-hidden="true">
      <div class="modal-dialog modal-lg">
        <div class="modal-content">
          <div class="modal-header bg-light">
            <h5 class="modal-title"><i class="fa fa-fw fa-pencil"></i> Edit <span class="text-primary">{{currentUser.login}}</span>'s profile</h5>
            <button type="button" class="close" data-dismiss="modal" aria-label="Close">
              <span aria-hidden="true">&times;</span>
            </button>
          </div>
          <div class="modal-body px-4 pt-3 pb-0">
            <user-form ref="user" :user="currentUser" :admin="true"></user-form>
          </div>
          <div class="modal-footer border-0 dbg-light">
            <button type="button" class="btn" data-dismiss="modal">Close</button>
            <button type="button" class="btn btn-success" @click.prevent="updateProfile">Save profile</button>
          </div>
        </div>
      </div>
    </div>

    <div class="modal fade email-user-modal" tabindex="-1" role="dialog" aria-labelledby="emailUserModal" aria-hidden="true">
      <div class="modal-dialog modal-lg">
        <div class="modal-content">
          <div class="modal-header bg-light">
            <h5 class="modal-title">
              <i class="fa fa-fw fa-envelope"></i>
              Send email to {{currentUser.firstName}} {{currentUser.lastName}} <span class="text-muted">({{ currentUser.login }})</span>
            </h5>
            <button type="button" class="close" data-dismiss="modal" aria-label="Close">
              <span aria-hidden="true">&times;</span>
            </button>
          </div>
          <div class="modal-body px-4 pt-3 pb-0">
            <email-form ref="mail" :to="currentUser.email + ', admin@iot-lab.info'"></email-form>
          </div>
          <div class="modal-footer border-0 dbg-light">
            <button type="button" class="btn" data-dismiss="modal">Close</button>
            <button type="button" class="btn btn-success" @click.prevent="$refs.mail.send">Send</button>
          </div>
        </div>
      </div>
    </div>

  </div> <!-- container -->
</template>

<script>
import UserForm from '@/components/UserForm'
import EmailForm from '@/components/EmailForm'
import {iotlab} from '@/rest'
import {auth} from '@/auth'
import $ from 'jquery'

export default {
  name: 'AdminUsers',

  components: {UserForm, EmailForm},

  data () {
    return {
      users: [],
      currentUser: {},
      auth: auth,
      pattern: '',
      show: 'pending',
    }
  },

  async created () {
    this.users = await iotlab.getUsers()
  },

  methods: {
    async showPending () {
      this.pattern = ''
      this.show = 'pending'
      this.users = await iotlab.getUsers({status: 'pending'})
    },
    async showAdmin () {
      this.pattern = ''
      this.show = 'admin'
      this.users = await iotlab.getUsers({isAdmin: true})
    },
    async search () {
      if (this.pattern) {
        this.show = false
        this.users = await iotlab.getUsers({search: this.pattern})
      } else {
        this.show = 'pending'
        this.showPending()
      }
    },
    async updateProfile () {
      if (await this.$refs.user.validate()) {
        try {
          await iotlab.setUserInfo(this.currentUser, this.currentUser.login)
          $('.edit-user-modal').modal('hide')
        } catch (err) {
          this.$notify({text: 'An error occured', type: 'error'})
        }
      }
    },
    async toggleActive (user) {
      if (user.status === 'pending') {
        if (confirm('Activate user?')) {
          try {
            await iotlab.activateUser(user.login)
            user.status = 'active'
            user.admin = false
          } catch (err) {
            this.$notify({text: 'An error occured', type: 'error'})
          }
        }
      } else {
        if (confirm('Deactivate user?')) {
          try {
            await iotlab.deactivateUser(user.login)
            user.status = 'pending'
            user.admin = false
          } catch (err) {
            this.$notify({text: 'An error occured', type: 'error'})
          }
        }
      }
    },
    async toggleAdmin (user) {
      if (confirm(`${user.admin ? 'Remove' : 'Set'} role admin for user?`)) {
        try {
          await iotlab.setAdminRole(user.login, !user.admin)
          user.admin = !user.admin
        } catch (err) {
          this.$notify({text: 'An error occured', type: 'error'})
        }
      }
    },
    async resetPassword (user) {
      if (confirm(`Reset password for user ${user.login}?\n\nThe user will receive a new password by email.`)) {
        try {
          await iotlab.resetPassword(user.email)
          this.$notify({text: 'New password sent by email', type: 'success'})
        } catch (err) {
          this.$notify({text: 'An error occured', type: 'error'})
        }
      }
    },
    async deleteUser (user) {
      if (confirm(`Delete user ${user.login}?`)) {
        try {
          await iotlab.deleteUser(user.login)
          this.users = this.users.filter(item => item.login !== user.login)
          this.$notify({text: `User ${user.login} deleted`, type: 'success'})
        } catch (err) {
          this.$notify({text: 'An error occured', type: 'error'})
        }
      }
    },
  },
}
</script>

<style>

</style>
