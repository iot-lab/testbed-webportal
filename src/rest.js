import axios from 'axios'

const API_URL = `https://${process.env.VUE_APP_IOTLAB_HOST}/api/`

export const iotlab = {

  api: axios.create({baseURL: API_URL, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')}),

  create (username, password) {
    this.api = axios.create({
      baseURL: API_URL,
      auth: {
        username: username,
        password: password,
      },
    })
  },

  // USERS API

  async signup (user, nbUsers) {
    if (nbUsers) {
      // create multiple user accounts
      await iotlab.api.post(`/users?nbusers=${nbUsers}`, user)
    } else {
      // create single user account
      await iotlab.api.post('/users', user)
    }
  },

  async activateAccount (hash) {
    await iotlab.api.post('/users/activate/', {'hash': hash})
  },

  async deleteUser (login = undefined) {
    if (login) {
      await iotlab.api.delete(`/users/${login}`)
    } else {
      await iotlab.api.delete(`/user`)
    }
  },

  async resetPassword (email) {
    await iotlab.api.post(`/users/reset_password`, {'email': email})
  },

  async changePassword (oldPwd, newPwd, confirmPwd) {
    await iotlab.api.put('/user/password', {
      'old_password': oldPwd,
      'new_password': newPwd,
      'confirm_new_password': confirmPwd,
    })
  },

  async getSSHkeys () {
    return iotlab.api.get('/user/keys').then(resp => resp.data.sshkeys)
  },

  async modifySSHkeys (keys) {
    return iotlab.api.post('/user/keys', {'sshkeys': keys})
  },

  async deleteSSHkey (id) {
    return iotlab.api.delete(`/user/keys/${id}`)
  },

  async getUsers ({status, group, search} = {}) {
    let params = {}
    if (status) {
      params.status = status
    }
    if (group) {
      params.group = group
    }
    if (search) {
      params.search = search
    }
    return iotlab.api.get('/users', {params: params}).then(resp => resp.data)
  },

  async getUserInfo (login = undefined) {
    if (login) {
      return iotlab.api.get(`/users/${login}`).then(resp => resp.data)
    } else {
      return iotlab.api.get('/user').then(resp => resp.data)
    }
  },

  async setUserInfo (user, login = undefined) {
    if (login) {
      return iotlab.api.put(`/users/${login}`, user)
    } else {
      return iotlab.api.put('/user', user)
    }
  },

  async activateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'active'})
  },

  async deactivateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'inactive'})
  },

  async setUserGroups (login, groups) {
    return iotlab.api.put(`/users/${login}`, {'groups': groups})
  },

  async sendMail (recipients, subject, message) {
    await iotlab.api.post('/users/email', {
      recipients: recipients,
      subject: subject,
      body: message,
    })
  },

  // EXPERIMENT API

  async getUserExperimentsCount () {
    return iotlab.api.get('/experiments/total').then(resp => resp.data)
  },

  async getAllExperimentsCount (user) {
    if (user) {
      return iotlab.api.get(`/experiments/total/users/${user}`).then(resp => resp.data)
    } else {
      return iotlab.api.get(`/experiments/total/all`).then(resp => resp.data)
    }
  },

  async getAllExperiments ({limit, offset, user, state = 'Terminated,Error,Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'} = {}) {
    let params = {}
    if (state) {
      params.state = state
    }
    if (limit) {
      params.limit = Math.min(limit, 500) // IoT-LAB api restrict queries to 500 entries
    }
    if (offset) {
      params.offset = offset
    }
    if (user === '@self') {
      return iotlab.api.get('/experiments', {params: params}).then(resp => resp.data.items)
    } else if (user) {
      return iotlab.api.get(`/experiments/users/${user}`, {params: params}).then(resp => resp.data.items)
    } else {
      return iotlab.api.get('/experiments/all', {params: params}).then(resp => resp.data.items)
    }
  },

  async getRunningExperiments () {
    return iotlab.api.get('/experiments/running').then(resp => resp.data.items)
  },

  async getExperiment (id) {
    return iotlab.api.get(`/experiments/${id}`).then(resp => resp.data)
  },

  async getExperimentArchive (id) {
    return iotlab.api.get(`/experiments/${id}/data`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async getExperimentDeployment (id) {
    return iotlab.api.get(`/experiments/${id}/deployment`).then(resp => resp.data)
  },

  async stopExperiment (id) {
    return iotlab.api.delete(`/experiments/${id}`)
  },

  async submitExperiment ({type, name, duration, reservation, nodes, firmwareassociations, firmwares, profileassociations, profiles} = {}) {
    const formData = new FormData()
    formData.append('experiment.json', JSON.stringify({
      type: type,
      duration: duration,
      reservation: reservation,
      name: name,
      nodes: nodes,
      firmwareassociations: firmwareassociations,
      profileassociations: profileassociations,
    }))
    // attached firmware names should be unique
    let attachedFirmwares = []
    for (let asso of firmwareassociations || []) {
      if (asso.firmwarename in firmwares && !attachedFirmwares.includes(asso.firmwarename)) {
        formData.append(asso.firmwarename, new Blob([firmwares[asso.firmwarename]], {type: 'application/octet-stream'}), asso.firmwarename)
        attachedFirmwares.push(asso.firmwarename)
      }
    }
    return iotlab.api.post('/experiments', formData).then(resp => resp.data)
  },

  async checkFirmware (fileString) {
    const formData = new FormData()
    formData.append('firmware.bin', fileString)
    return iotlab.api.post('/firmwares/checker', formData).then(resp => resp.data)
  },

  async flashFirmware (id, nodes, fileString) {
    const formData = new FormData()
    formData.append(`experiment${id}.json`, JSON.stringify(nodes))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/experiments/${id}/nodes/flash`, formData).then(resp => resp.data)
  },

  async flashResourcesFirmware (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/flash/${name}`, nodes).then(resp => resp.data)
  },

  async updateMonitoring (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/monitoring/${name}`, nodes).then(resp => resp.data)
  },

  async updateExperimentMobilityCircuit (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/robots/mobility/${name}`, nodes).then(resp => resp.data)
  },

  async reloadExperiment (id) {
    return iotlab.api.post(`/experiments/${id}/reload`, {}).then(resp => resp.data)
  },

  // USER GROUPS

  async getUserGroups () {
    return iotlab.api.get('/groups').then(resp => resp.data)
  },

  async createGroup (name) {
    return iotlab.api.post(`/groups`, {name: name})
  },

  async deleteGroup (group) {
    return iotlab.api.delete(`/groups/${group}`)
  },

  // OTHER API

  async getSites () {
    return iotlab.api.get('/sites').then(resp => resp.data.items)
  },

  async getSitesDetails () {
    return iotlab.api.get('/sites/details').then(resp => resp.data.items)
  },

  async getNodes () {
    return iotlab.api.get('/nodes').then(resp => resp.data.items)
  },

  async updateNodesProperties () {
    return iotlab.api.put('/nodes')
  },

  async updateNodesPropertiesOldApi () {
    const OLD_API = `https://${process.env.VUE_APP_IOTLAB_HOST}/rest/`
    const oldApi = axios.create({baseURL: OLD_API, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')})
    return oldApi.post('/admin/nodes')
  },

  async getNodesIds () {
    return iotlab.api.get('/nodes/ids').then(resp => resp.data.items)
  },

  async getExperimentNodes (id) {
    return iotlab.api.get(`/experiments/${id}/nodes`).then(resp => resp.data.items)
  },

  async getExperimentToken (id) {
    return iotlab.api.get(`/experiments/${id}/token`).then(resp => resp.data.token)
  },

  async sendNodesCommand (id, cmd, nodes) {
    // cmd in [start, stop, reset, flash-idle, profile-reset, debug-start, debug-stop]
    return iotlab.api.post(`/experiments/${id}/nodes/${cmd}`, nodes).then(resp => resp.data)
  },

  // STORES (firmwares, profiles, etc)

  async getMobilityCircuits () {
    return iotlab.api.get('/mobilities/circuits').then(resp => resp.data.items)
  },

  async getMobilityCircuit (name) {
    return iotlab.api.get(`/mobilities/circuits/${name}`).then(resp => resp.data)
  },

  async getRobotsSiteMapImage (site) {
    return iotlab.api.get(`/robots/${site}/map/image`, {
      responseType: 'arraybuffer',
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    }).then(resp => Buffer.from(resp.data, 'binary').toString('base64')
    )
  },

  async getRobotsSiteMapConfig (site) {
    return iotlab.api.get(`/robots/${site}/map/config`).then(resp => resp.data)
  },

  async createMobilityCircuit (circuit) {
    return iotlab.api.post(`/mobilities/circuits`, circuit)
  },

  async updateMobilityCircuit (name, circuit) {
    return iotlab.api.put(`/mobilities/circuits/${name}`, circuit)
  },

  async deleteMobilityCircuit (name) {
    return iotlab.api.delete(`/mobilities/circuits/${name}`)
  },

  async getMonitoringProfiles () {
    return iotlab.api.get('/monitoring').then(resp => resp.data)
  },

  async getMonitoringProfile (name) {
    return iotlab.api.get(`/monitoring/${name}`).then(resp => resp.data)
  },

  async updateMonitoringProfile (name, profile) {
    return iotlab.api.put(`/monitoring/${name}`, profile)
  },

  async createMonitoringProfile (profile) {
    return iotlab.api.post(`/monitoring`, profile)
  },

  async deleteMonitoringProfile (name) {
    return iotlab.api.delete(`/monitoring/${name}`)
  },

  async getFirmwares () {
    return iotlab.api.get('/firmwares').then(resp => resp.data)
  },

  async getFirmware (name) {
    return iotlab.api.get(`/firmwares/${name}`).then(resp => resp.data)
  },

  async getFirmwareFile (name) {
    return iotlab.api.get(`/firmwares/${name}/file`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async updateFirmware (name, firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    if (fileString) {
      formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    }
    return iotlab.api.put(`/firmwares/${name}`, formData).then(resp => resp.data)
  },

  async createFirmware (firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/firmwares`, formData).then(resp => resp.data)
  },

  async deleteFirmware (name) {
    return iotlab.api.delete(`/firmwares/${name}`)
  },
}
