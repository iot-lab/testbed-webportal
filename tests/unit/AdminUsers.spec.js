import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import AdminUsers from '@/views/AdminUsers'
import { iotlab } from '@/rest'
jest.mock('@/rest')

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter()

describe('AdminUsers.vue', () => {

  it('should match snapshot', () => {
    const wrapper = shallowMount(AdminUsers, {localVue, router})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
