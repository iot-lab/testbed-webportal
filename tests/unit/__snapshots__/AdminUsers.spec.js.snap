// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminUsers.vue should match snapshot 1`] = `
<div
  class="container mt-3"
>
  <h4
    class="d-none d-sm-block d-md-none"
  >
    Users
  </h4>
   
  <form
    class="form-inline my-4"
  >
    <div
      class="input-group mr-2"
    >
      <input
        class="form-control"
        placeholder="Search for users"
        type="text"
      />
       
      <div
        class="input-group-btn"
      >
        <button
          aria-label="Search"
          class="btn btn-secondary"
          type="submit"
        >
          <i
            class="fa fa-search"
          />
        </button>
      </div>
    </div>
     
    <select
      class="form-control custom-select mr-2"
    >
      <option
        disabled="disabled"
        value=""
      >
        Show user group
      </option>
       
      <option
        value="inactive"
      >
        inactive users
      </option>
       
    </select>
     
    <i
      class="fa fa-lg fa-spinner fa-spin ml-1 mr-auto text-white"
    />
     
    <router-link-stub
      class="btn btn-success mr-2"
      event="click"
      tag="a"
      to="[object Object]"
    >
      <i
        class="fa fa-user-plus"
      />
       Add Users
    </router-link-stub>
     
    <router-link-stub
      class="btn btn-secondary mr-2"
      event="click"
      tag="a"
      to="[object Object]"
    >
      <i
        class="fa fa-cog"
      />
       Groups
    </router-link-stub>
     
    <div
      class="dropdown d-inline-block"
    >
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class="btn dropdown-toggle cursor"
        data-toggle="dropdown"
        id="actionMenuBtn"
        type="button"
      >
        <i
          class="fa fa-wrench text-dark"
        />
         Actions on  users
      
      </button>
       
      <div
        aria-labelledby="actionMenuBtn"
        class="dropdown-menu dropdown-menu-right"
      >
        <a
          class="dropdown-item"
          href="#"
        >
          <i
            class="fa fa-fw fa-user-plus"
          />
           Add groups
        </a>
         
        <a
          class="dropdown-item"
          href="#"
        >
          <i
            class="fa fa-fw fa-user-times"
          />
           Remove groups
        </a>
         
        <a
          class="dropdown-item text-danger"
          href="#"
        >
          <i
            class="fa fa-fw fa-trash"
          />
           Delete users
        </a>
      </div>
    </div>
  </form>
   
  <table
    class="table table-striped table-sm"
  >
    <thead>
      <tr>
        <th
          class="header"
          width="15px"
        >
          <input
            type="checkbox"
          />
        </th>
         
        <th
          aria-sort="ascending"
          class="header headerSortDown"
        >
          Login
        </th>
         
        <th
          class="header"
        >
          Name
        </th>
         
        <th
          class="header"
        >
          Email
        </th>
         
        <th
          class="header"
        >
          Created
        </th>
         
        <th
          class="header"
          width="50px"
        >
          State
        </th>
         
        <th
          class="header"
        >
          Groups
        </th>
         
        <th
          class="header"
          width="50px"
        >
          Actions
        </th>
      </tr>
    </thead>
     
    <tbody />
  </table>
   
  <div
    class="text-muted"
  >
    <p>
      Select a search criterion
    </p>
  </div>
   
  <modaldialog-stub
    name="add-groups-modal"
  >
    <template>
      <i
        class="fa fa-fw fa-user-plus"
      />
       Add groups to 
      <span
        class="text-primary"
      >
        0 users
      </span>
    </template>
      
    <template>
      Add groups
    </template>
  </modaldialog-stub>
   
  <modaldialog-stub
    name="remove-groups-modal"
  >
    <template>
      <i
        class="fa fa-fw fa-user-plus"
      />
       Remove groups to 
      <span
        class="text-primary"
      >
        0 users
      </span>
    </template>
      
    <template>
      Remove groups
    </template>
  </modaldialog-stub>
   
  <modaldialog-stub
    name="edit-groups-modal"
  >
    <template>
      <i
        class="fa fa-fw fa-pencil"
      />
       Edit user groups for 
      <span
        class="text-primary"
      >
        
      </span>
    </template>
     
  </modaldialog-stub>
   
  <div
    aria-hidden="true"
    aria-labelledby="editUserModal"
    class="modal fade edit-user-modal"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-lg"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header bg-light"
        >
          <h5
            class="modal-title"
          >
            <i
              class="fa fa-fw fa-pencil"
            />
             Edit 
            <span
              class="text-primary"
            >
              
            </span>
            's profile
          </h5>
           
          <button
            aria-label="Close"
            class="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
         
        <div
          class="modal-body px-4 pt-3 pb-0"
        >
          <userform-stub
            admin="true"
            hidden="password"
            mode="edit"
            user="[object Object]"
          />
        </div>
         
        <div
          class="modal-footer border-0 dbg-light"
        >
          <button
            class="btn"
            data-dismiss="modal"
            type="button"
          >
            Close
          </button>
           
          <button
            class="btn btn-success"
            type="button"
          >
            Save profile
          </button>
        </div>
      </div>
    </div>
  </div>
   
  <div
    aria-hidden="true"
    aria-labelledby="emailUserModal"
    class="modal fade email-user-modal"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-lg"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header bg-light"
        >
          <h5
            class="modal-title"
          >
            <i
              class="fa fa-fw fa-envelope"
            />
            
            Send email to   
            <span
              class="text-muted"
            >
              ()
            </span>
          </h5>
           
          <button
            aria-label="Close"
            class="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
         
        <div
          class="modal-body px-4 pt-3 pb-0"
        >
          <emailform-stub
            message="Dear IoT-LAB user,



Regards,
The FIT IoT-LAB Team
https://www.iot-lab.info"
            subject=""
            to="undefined, admin@iot-lab.info"
          />
        </div>
         
        <div
          class="modal-footer border-0"
        >
          <button
            class="btn"
            data-dismiss="modal"
            type="button"
          >
            Close
          </button>
           
          <button
            class="btn btn-success"
            type="button"
          >
            Send
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
